File: src/main/java/edu/cornell/cs/apl/viaduct/protocol/ProtocolSelection.java
package edu.cornell.cs.apl.viaduct.protocol;

import edu.cornell.cs.apl.viaduct.AstNode;
import edu.cornell.cs.apl.viaduct.imp.HostTrustConfiguration;
import edu.cornell.cs.apl.viaduct.pdg.PdgNode;
import edu.cornell.cs.apl.viaduct.pdg.ProgramDependencyGraph;
import io.vavr.collection.Map;

public interface ProtocolSelection<T extends AstNode> {
  Map<PdgNode<T>, Protocol<T>> selectProtocols(
      HostTrustConfiguration hostConfig, ProgramDependencyGraph<T> pdg);
}

File: src/main/kotlin/edu/cornell/cs/apl/viaduct/selection/ProtocolSelection.kt
package edu.cornell.cs.apl.viaduct.selection

import edu.cornell.cs.apl.viaduct.syntax.Protocol
import edu.cornell.cs.apl.viaduct.syntax.Variable
import edu.cornell.cs.apl.viaduct.syntax.intermediate.StatementNode
import kotlinx.collections.immutable.PersistentMap

typealias ProtocolMap = PersistentMap<Variable, Protocol>

interface ProtocolSelection {
    fun selectProtocols(context: ProtocolSelectionContext, stmt: StatementNode): ProtocolMap?
}
